zap_otlp_sync/zap_otlp_sync.go

```diff
@@ -0,0 +1,182 @@
+	BatchSize      int
+	BatchInterval  time.Duration
+	ResourceSchema string
+	Scope          *instrumentation.Scope
+	Resource       *resource.Resource
+}
+
+func NewOtlpSyncer(conn *grpc.ClientConn, options Options) *OtelSyncer {
+	ctx, cancel := context.WithCancel(context.Background())
+
+	if options.BatchSize == 0 {
+		options.BatchSize = 100
+	}
+
+	if options.BatchInterval == 0 {
+		options.BatchInterval = time.Duration(5) * time.Second
+	}
+
+	var rattrs *rv1.Resource
+	if options.Resource != nil {
+		iter := options.Resource.Iter()
+		attrs := make([]*cv1.KeyValue, 0, iter.Len())
+		for iter.Next() {
+			attr := iter.Attribute()
+			attrs = append(attrs, &cv1.KeyValue{
+				Key:   string(attr.Key),
+				Value: &cv1.AnyValue{Value: &cv1.AnyValue_StringValue{StringValue: attr.Value.Emit()}},
+			})
+		}
+		rattrs = &rv1.Resource{Attributes: attrs}
+	}
+
+	syncer := &OtelSyncer{
+		ctx:        ctx,
+		close:      cancel,
+		res:        rattrs,
+		resSchema:  options.ResourceSchema,
+		values:     [][]byte{},
+		queue:      make(chan []byte, options.BatchSize), // keeping queue size as batch size
+		client:     collpb.NewLogsServiceClient(conn),
+		batchSize:  options.BatchSize,
+		sendBatch:  make(chan bool),
+		valueMutex: sync.Mutex{},
+		pushDataWg: sync.WaitGroup{},
+	}
+
+	go syncer.processQueue(options.BatchInterval)
+
+	return syncer
+}
+
+func (l *OtelSyncer) processQueue(interval time.Duration) {
+	ticker := time.NewTicker(interval)
+	defer ticker.Stop()
+	for {
+		select {
+		case <-l.ctx.Done():
+			return
+		case <-ticker.C:
+			l.pushData()
+		case data := <-l.queue:
+			l.valueMutex.Lock()
+			l.values = append(l.values, data)
+			shouldExport := len(l.values) >= l.batchSize
+			l.valueMutex.Unlock()
+			if shouldExport {
+				l.pushData()
+			}
+		case <-l.sendBatch:
+			l.pushData()
+		}
+	}
+}
+
+func (l *OtelSyncer) pushData() (err error) {
+	l.pushDataWg.Add(1)
+	defer l.pushDataWg.Done()
+
+	rec := map[string][]*lpb.LogRecord{}
+
+	l.valueMutex.Lock()
+	for _, v := range l.values {
+		data := strings.Split(string(v), "#SIGNOZ#")
+		r := &lpb.LogRecord{}
+		err = proto.Unmarshal([]byte(data[1]), r)
+		if err != nil {
+			l.valueMutex.Unlock()
+			return err
+		}
+		rec[data[0]] = append(rec[data[0]], r)
+	}
+	l.values = [][]byte{} // clean the values
+	l.valueMutex.Unlock()
+
+	if len(rec) == 0 {
+		return nil
+	}
+
+	rl := &lpb.ResourceLogs{ScopeLogs: []*lpb.ScopeLogs{}}
+	for logger, records := range rec {
+		sl := &lpb.ScopeLogs{LogRecords: records}
+		if logger != "" {
+			sl.SchemaUrl, sl.Scope = l.resSchema, &cv1.InstrumentationScope{
+				Name:    "logger",
+				Version: logger,
+			}
+		}
+		rl.ScopeLogs = append(rl.ScopeLogs, sl)
+	}
+
+	if l.res != nil {
+		rl.SchemaUrl, rl.Resource = l.resSchema, l.res
+	}
+	_, err = l.client.Export(context.Background(), &collpb.ExportLogsServiceRequest{
+		ResourceLogs: []*lpb.ResourceLogs{rl},
+	})
+	// TODO: how to handle partial failure and error
+	if err != nil {
```

Comment:
Should we also log the error here?

Reply:
we can use a fmt.Errorf here I guess, but not totally sure.

Reply:
Let's log it, I have seen exporters logging errors.

Reply:
Curious why `fmt`, not `log`?

Reply:
Yah log will be better, it will include ts as well. Let me update it.